Feature/ homework 5

VtulkinVladimir opened the pull request:
переделан прошлый проект из дз с учетом архитектуры

VtulkinVladimir committed:
Дублировал проект с прошлой домашки

три экрана. первый с коллекшн вью, второй с детальными данными и третий, открывающийся модально

VtulkinVladimir committed:
делегаты

VtulkinVladimir committed:
первый экран MVC

VtulkinVladimir committed:
второй экран MVP

VtulkinVladimir committed:
третий экран MVVM

VtulkinVladimir committed:
добавил проект в воркспейс

UdmurTio commented on CollectionApp with architecture/CollectionApp/CarModel.swift:
```diff
@@ -0,0 +1,90 @@
+
+struct CarModel
+{
+	static let data = [
```
Смотри. У тебя модель - структура, значит везде, куда ее передаешь - будет копия, а нам копия источника данных не нужна.

UdmurTio commented on CollectionApp with architecture/CollectionApp/CarModel.swift:
```diff
@@ -0,0 +1,90 @@
+import UIKit
+
+struct CarModel
+{
```
Смотри. У тебя модель - структура, значит везде, куда ее передаешь - будет копия, а нам копия источника данных не нужна.

UdmurTio commented on CollectionApp with architecture/CollectionApp/CarModel.swift:
```diff
@@ -0,0 +1,90 @@
+struct CarModel
+{
+	static let data = [
+		CarModel(image: UIImage(named: "HondaCivic6"),
```
Также у тебя модель является и самими данными и источником данных - нехорошо. Лучше, чтоб был класс с методами там get set или load save, внутри него ты в каком-то формате хранишь/получаешь данные, а наружу они, в виде структуры, как раз выходят по методам класса

UdmurTio commented on CollectionApp with architecture/CollectionApp/ExchengeRates.swift:
```diff
@@ -0,0 +1,12 @@
+struct ExchengeRates
+{
+	static let usdToRub: Double = 70
+	static let usdToEuro: Double = 0.89
```
В 99% случаев Float достаточно (это просто на будущее)

UdmurTio commented on CollectionApp with architecture/CollectionApp/FirstScreen(MVC)/CarCollectionCell.swift:
```diff
@@ -0,0 +1,97 @@
+	private let model = UILabel()
+	private let yearOfIssue = UILabel()
+	private let fontSize = TextSize.carCollectionCellTextSize
+
```
В идеале вспомогательные данные выносить в отдельную приватную структурку с набором статических констант и как-нибудь их называть. А то у тебя получается, что свойством ячейки является fontSize, что странно, ведь fontSize тебе нужен только для данных внутри лейблов

UdmurTio commented on CollectionApp with architecture/CollectionApp/FirstScreen(MVC)/MainScreenCarModel.swift:
```diff
@@ -0,0 +1,22 @@
+	let model: String
+	let yearOfIssue: Int
+	
+	init(car: CarModel) {
```
Также проблема - вроде моделька, которая должна выдавать/получать данные, а вроде одновременно и сам класс хранения этих данных

UdmurTio commented on CollectionApp with architecture/CollectionApp/FirstScreen(MVC)/MainScreenView.swift:
```diff
@@ -0,0 +1,66 @@
+final class MainScreenView: UIView
+{
+	var didTapHandler: ((Int)->Void)?
+	private var cars: [MainScreenCarModel]?
```
В названии класса написано, что это модель, значит получается, что вьюха знает о модели - надо подправить, у нее на вход должна поступать какая-то нибудь ViewModelData или что-то в таком роде

UdmurTio commented on CollectionApp with architecture/CollectionApp/FirstScreen(MVC)/MainScreenView.swift:
```diff
@@ -0,0 +1,66 @@
+
+final class MainScreenView: UIView
+{
+	var didTapHandler: ((Int)->Void)?
```
Странно, вроде хендлер, сигнализирующий о нажатии, а вроде в него передается какое-то целое число

UdmurTio commented on CollectionApp with architecture/CollectionApp/FirstScreen(MVC)/MainScreenView.swift:
```diff
@@ -0,0 +1,66 @@
+{
+	func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+		let cell = collectionView.dequeueReusableCell(withReuseIdentifier: CarCollectionCell.identifier, for: indexPath) as! CarCollectionCell
+		cell.car = self.cars?[indexPath.row]
```
Здесь желательно вызывать какой-нибудь хендлер, который будет задан из контроллера и который будет возвращать тебе данные, а то получается, что вью сама решает, что и когда ей показать

UdmurTio commented on CollectionApp with architecture/CollectionApp/FirstScreen(MVC)/MainScreenView.swift:
```diff
@@ -0,0 +1,66 @@
+	}
+	
+	func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+		return self.cars?.count ?? 0
```
То же самое

UdmurTio commented on CollectionApp with architecture/CollectionApp/FirstScreen(MVC)/MainScreenView.swift:
```diff
@@ -0,0 +1,66 @@
+extension MainScreenView: UICollectionViewDelegate
+{
+	func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+		self.didTapHandler?(indexPath.row)
```
А вот здесь хорошо сделано, через хендлер как раз, только надо переименовать, а то не понятно, вроде просто тапХендлер, а вроде в него передается индексПатх. Лучше, чтобы он назывался как-то типа didTapCellHandler

UdmurTio commented on CollectionApp with architecture/CollectionApp/FirstScreen(MVC)/MainScreenViewController.swift:
```diff
@@ -0,0 +1,45 @@
+
+final class MainScreenViewController: UIViewController
+{
+	private var cars = CarModel.data
```
Тут должны быть ссылка просто на модель, а не уже на готовые данные. И где надо эти данные с модели запрашивать

UdmurTio commented on CollectionApp with architecture/CollectionApp/FirstScreen(MVC)/CarCollectionCell.swift:
```diff
@@ -0,0 +1,97 @@
+	private let fontSize = TextSize.carCollectionCellTextSize
+
+	
+	var car: MainScreenCarModel? {
```
Данные лучше проставлять не напрямую, а через какой-нибудь метод торчащий наружу типа update или вообще разбивать на методы типа setCar showCar. А то получается, что мы вроде выставляем что-то для вьюхи, а по факту вью обновляется (показываются всякие там тени, меняются шрифты и т.д.) - немного неожиданное поведение с точки зрения того, кто просто выставил данные

UdmurTio commented on CollectionApp with architecture/CollectionApp/SecondScreen(MVP)/CarDetailPresenter.swift:
```diff
@@ -0,0 +1,36 @@
+//  Created by Владимир Втулкин on 29.11.2021.
+//
+
+final class CarDetailPresenter
```
Хоть презентер и редко переиспользуется - все рано лучше закрывать протоколом, раз уж вью закрываешь

UdmurTio commented on CollectionApp with architecture/CollectionApp/SecondScreen(MVP)/CarDetailView.swift:
```diff
@@ -0,0 +1,100 @@
+	private let infoButton = UIButton(type: .infoLight)
+	private var screenSize = UIScreen.main.bounds
+
+	private func configureView() {
```
Лучше разбить на пару приватных методов вроде addSubviews, configSubviews и makeConstraintsSubviews

UdmurTio commented on CollectionApp with architecture/CollectionApp/SecondScreen(MVP)/CarDetailViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+import UIKit
+
+final class CarDetailViewController: UIViewController
```
Тогда уже и контроллер закрывай протоколом))

UdmurTio commented on CollectionApp with architecture/CollectionApp/SecondScreen(MVP)/CarDetailView.swift:
```diff
@@ -0,0 +1,100 @@
+
+extension CarDetailView: ICarDetailView
+{
+	func setContent(car: CarModel) {
```
Опять одна и та же модель для разных слоев

UdmurTio commented on CollectionApp with architecture/CollectionApp/ThirtScreen(MVVM)/PriceViewModel.swift:
```diff
@@ -0,0 +1,38 @@
+//
+//  Created by Владимир Втулкин on 01.12.2021.
+//
+final class ListnersManager
```
Слов Manager, Service, Helper и т.д. лучше избегать, так как они дают не полное понимание о том, чем объект должен заниматься

UdmurTio reviewed (commented):
Работа неплохая, но нужно немножко подправить

VtulkinVladimir committed:
поправил первый модуль

исправил работу с моделями и хранение данных. убрал с вью хранение, переделал на хэндлеры

VtulkinVladimir committed:
поправил второй модуль

закрыл вью контроллер и презентер протоколами. переделал модель. во вью разбил configureView на несколько более мелких методов

VtulkinVladimir committed:
поправил третий модуль

фантазии на переименовывание хватило только на это)))

VtulkinVladimir commented:
самое интересное, что большую часть из этих косяков я уже осознал самостоятельно пока делал новую домашку, и не допустил их. ну по крайней мере мне так кажется)) там будут наверняка свои, особенные
